### PR TITLE
Properly marshal optional int

### LINKED ISSUE #905: Crash when creating a save with RealSolarSystem on Buffon
Reported by Ice__.
INFO file: https://gist.github.com/anonymous/8394ed2257abcf1052d0.
```
F0223 17:14:36.820339 13212 parser_body.hpp:111] Unsupported unit of angle
```
Decoded stack trace:
[`quantities/parser_body.hpp:111`](https://github.com/mockingbirdnest/Principia/blob/0455d0cb0e4b0b584a84caf40520cf7993903e0c/quantities/parser_body.hpp#L111)
[`quantities/parser_body.hpp:74`](https://github.com/mockingbirdnest/Principia/blob/0455d0cb0e4b0b584a84caf40520cf7993903e0c/quantities/parser_body.hpp#L74)
[`physics/solar_system_body.hpp:207`](https://github.com/mockingbirdnest/Principia/blob/0455d0cb0e4b0b584a84caf40520cf7993903e0c/physics/solar_system_body.hpp#L207)
[`ksp_plugin/interface.cpp:111`](https://github.com/mockingbirdnest/Principia/blob/0455d0cb0e4b0b584a84caf40520cf7993903e0c/ksp_plugin/interface.cpp#L111)
[`ksp_plugin/interface.cpp:356`](https://github.com/mockingbirdnest/Principia/blob/0455d0cb0e4b0b584a84caf40520cf7993903e0c/ksp_plugin/interface.cpp#L356)

### PR DESCRIPTION
**Pull request by eggrobin**
Fix #905 as well as a typo (`mean_value` instead of `mean_radius`).
See the [changes to the generated code](https://github.com/mockingbirdnest/Principia/commit/ef88c0ebc33ba37a09ba3d0ca126e26de51ba09f).

### CODE REVIEW

File: ksp_plugin_adapter/ksp_plugin_adapter.cs [hunk 1]
```diff
@@ -1313,44 +1313,20 @@ where is_in_physics_bubble(vessel)
...
-                sun_gravity_model.HasValue("reference_radius") ?
-                sun_gravity_model.GetValue("reference_radius") : null,
-            x: sun_initial_state.GetValue("x"),
-            y: sun_initial_state.GetValue("y"),
-            z: sun_initial_state.GetValue("z"),
-            vx: sun_initial_state.GetValue("vx"),
-            vy: sun_initial_state.GetValue("vy"),
-            vz: sun_initial_state.GetValue("vz"));
         BodyProcessor insert_body = body => {
-          Log.Info("Inserting " + body.name + "...");
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: You don't want to keep this?

File: ksp_plugin_adapter/ksp_plugin_adapter.cs [hunk 2]
```diff
@@ -1313,44 +1313,20 @@ where is_in_physics_bubble(vessel)
...
+            parent_index = body.orbit.referenceBody.flightGlobalsIndex;
+          }
           plugin_.InsertCelestialAbsoluteCartesian(
               celestial_index: body.flightGlobalsIndex,
-              parent_index: ref parent_index,
+              parent_index: parent_index,
               gravitational_parameter:
                   gravity_model.GetValue("gravitational_parameter"),
               mean_radius:
-                  gravity_model.GetValue("mean_value"),
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: Oh.

File: ksp_plugin_adapter/optional_marshaler.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+public class OptionalMarshaler<T> : ICustomMarshaler where T : struct {
+  // In addition to implementing the |ICustomMarshaler| interface, custom
+  // marshalers must implement a static method called |GetInstance| that accepts
+  // a |String| as a parameter and has a return type of |ICustomMarshaler|,
+  // see https://goo.gl/wwmBTa.
+  public static ICustomMarshaler GetInstance(String s) {
+    return instance_;
+  }
+
+  public void CleanUpManagedData(object ManagedObj) {}
```
Comment (pleroy) [inline ksp_plugin_adapter/optional_marshaler.cs]: managed_objects

File: ksp_plugin_adapter/optional_marshaler.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  public void CleanUpNativeData(IntPtr pNativeData) {
```
Comment (pleroy) [inline ksp_plugin_adapter/optional_marshaler.cs]: native_data (just say nem to Hungarian).

File: ksp_plugin_adapter/optional_marshaler.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    Marshal.FreeHGlobal(pNativeData);
+  }
+
+  public int GetNativeDataSize() {
+    // I think this is supposed to return -1, and I also think it doesn't
+    // matter, but honestly I'm not sure...
+    return -1;
+  }
+
+  public IntPtr MarshalManagedToNative(object ManagedObj) {
```
Comment (pleroy) [inline ksp_plugin_adapter/optional_marshaler.cs]: managed_object

File: ksp_plugin_adapter/optional_marshaler.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    // While we expect that the marshaling attribute will be used on a |T?|,
+    // we get it boxed, and boxing has special behaviour on |Nullable|;
+    // specifically, |T?| is boxed to either |T| or |null|, depending on whether
+    // it has a value.  In the latter case, we lose the type information, so we
+    // cannot test whether |object is T?|.  Instead we check whether
+    // |object == null|, if it it's not, we check that it's a |T|.
```
Comment (pleroy) [inline ksp_plugin_adapter/optional_marshaler.cs]: if it is not

File: ksp_plugin_adapter/optional_marshaler.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+              typeof(T?).Name,
+              typeof(T).Name));
+    }
+    T value = value_if_correct_type.Value;
+    IntPtr ptr = Marshal.AllocHGlobal(Marshal.SizeOf(value));
+    Marshal.StructureToPtr(value, ptr, fDeleteOld: false);
+    return ptr;
+  }
+
+  public object MarshalNativeToManaged(IntPtr pNativeData) {
```
Comment (pleroy) [inline ksp_plugin_adapter/optional_marshaler.cs]: native_data

File: ksp_plugin_adapter/ksp_plugin_adapter.cs (hunk 1 above)
Comment (eggrobin) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: Oops.

File: ksp_plugin_adapter/optional_marshaler.cs (hunk 7 above)
Comment (eggrobin) [inline ksp_plugin_adapter/optional_marshaler.cs]: Done.

File: ksp_plugin_adapter/optional_marshaler.cs (hunk 6 above)
Comment (eggrobin) [inline ksp_plugin_adapter/optional_marshaler.cs]: Done.

File: ksp_plugin_adapter/optional_marshaler.cs (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin_adapter/optional_marshaler.cs]: Done.

File: ksp_plugin_adapter/optional_marshaler.cs (hunk 4 above)
Comment (eggrobin) [inline ksp_plugin_adapter/optional_marshaler.cs]: Done.

File: ksp_plugin_adapter/optional_marshaler.cs (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin_adapter/optional_marshaler.cs]: Done.

**Pull request merged**